LOG-15414 - Fix race condition that results in dropped logs

shawkinsmezmo opened the pull request:
This PR addresses the underlying issue uncovered by @aholmberg in LOG-15414. It makes the following modifications to address potential race scenarios:
* Ensure that the lock synchronizing access to the buffer is always released by adding a `try...finally` to the `buffer_log()` method. Previously, it was possible for this lock to not be released if an exception was thrown within `close_flusher()`, `flush()`, or `start_flusher()`. (The likelihood of this happening currently is low since `close_flusher()` and `start_flusher()` do not appear to do anything that could result in an exception, and `flush()` only adds things to a ThreadPoolExecutor, but this change more strongly enforces good behavior in the event that any of these methods are modified.
* Rather than attempting to acquire the lock prior to submitting `try_request()` to the request thread pool, attempt to acquire it from within the worker thread. The previous behavior would not properly protect access to the buffer, because the lock would be released after the task was submitted to the thread pool, but _not_ actually after the work in `try_request()` was completed, which clears the buffer upon completion.
* Re-order the operations in the `close()` method to ensure that the contents of the buffer can properly and safely drain during logger / application shutdown.
* Unit test that re-produces the underlying issue with the previous code, and passes with the above changes.

gjanco reviewed (dismissed):
Looks good to me.  But please wait for one of the SWEs to look at it too.

dkhokhlov commented on logdna/logdna.py:
```diff
@@ -135,28 +141,34 @@ def flush(self):
-    def flush_sync(self):
-        if self.buf_size == 0 and len(self.secondary) == 0:
-            return
+    def flush_sync(self, should_block=False):
```
"sync" in name implies blocking? isn't? 
why it is optional now?

shawkinsmezmo replied:
Previously, it was never blocking. The reason it is blocking now is because it is actually invoked in two contexts, one of which was previously subject to a race:
* When the flusher Timer thread wakes up
* When the size of the buffered data meets or exceeds the configured buffer size send threshold
* When the logger is being closed. This is where a race was identified where the request threads would be shut down while the worker threads were adding data to the buffer.  In this case, we want the final shutdown flush to block and wait to acquire the buffer lock so we can be sure to do a final drain and not lose any logged data.

Happy to consider renaming this method if anyone has any better suggestions.

dkhokhlov replied:
if flush_sync() was not blocking then it is a bug.

shawkinsmezmo replied:
I actually don't think it needs to (or should) be blocking in all circumstances. Because the method can get invoked in separate threads, there's not really a benefit to one thread waiting to acquire the lock when another is already active trying to drain the buffer. Whoever gets the lock first can complete the sync and re-initialize the buffer. The only reason we want it to block when waiting to acquire the lock during shutdown is to avoid a race where the lock isn't released quickly enough by other threads during shutdown.

dkhokhlov replied:
so the fix is for app shutdown case only?
what is the meaning of the "sync" here?

shawkinsmezmo replied:
@dkhokhlov More specifically, there were a few different races that were identified and addressed with this PR. The change to make the behavior here blocking for this specific scenario addresses one of those races.

As far as the term "sync", I can see how sync would imply "synchronized" or "synchronous", but I actually interpreted it differently when I first came across it and read the code, since it's intended purposes is to "sync" the data from the client to the server.  I'm not opinionated here on naming though and am happy to change it if you and others feel strongly, but I think that the actual meat and potatoes of this method is now doing what (in theory) it should be.

dkhokhlov replied:
> few different races 
the test is for shutdown only?

shawkinsmezmo replied:
The test covers all of the potential race scenarios that I uncovered. The main race is exposed when the logger is invoked at a rate sufficiently fast enough (which is not all that fast in actuality). In that scenario, the buffer would be drained while it was being written to by other threads. The other race is that one we were discussing previously during shutdown. The test covers the former case by sending a high volume of logs (10k) as quickly as possible, stressing the access synchronization to the buffer, and covers the latter case by shutting down the logger at the end and then subsequently verifying that all expected data was emitted by the logger.

dkhokhlov reviewed (dismissed):
lgtm

emelski commented on logdna/logdna.py:
```diff
@@ -102,21 +103,26 @@ def buffer_log_sync(self, message):
+            finally:
+                self.lock.release()
         else:
             self.secondary.append(message)
```
It seems like this is still quite vulnerable to a race condition, as the attempt to append to `self.secondary` is not protected by any lock here.

shawkinsmezmo replied:
@emelski Great catch and you are absolutely right. The idea of using a secondary buffer just strikes me as turtles all the way down for this type of application. It would probably be better to refactor the code to use a simple queue, which is already guaranteed to be thread-safe, rather than the current design of using a lock to synchronize access to a simple list.  Using a queue, we could remove the locking alltogether.

shawkinsmezmo replied:
To be clear, there is still locking happening behind the scenes, but it's encapsulated within the queue implementation itself, so at the very least it simplifies the code. Not sure if there will be any negative performance regressions with that sort of implementation so I will test it thoroughly.

emelski replied:
I'm not sure the built-in Queue has the right semantics for the use here -- that is, that Queue is more-or-less designed around the notion of each item in the queue being handled discretely from the others, whereas in our use we want to treat _everything currently in the queue_ as a single block of data to reduce the number of separate requests that are sent to our servers.  Obviously you could implement such semantics using the Queue but it's not quite as simple as saying "let's just use a queue".  In some ways it might be quite clunky to do, in fact, since the Queue would not prohibit threads adding to the queue at just the moment the flushing thread was trying to pull things out to buffer for transmission.  Again, not an unsolvable problem but not an obvious slam dunk replacement either, imo.

Also a minor semantic nit: it's misleading to suggest that using Queue would remove locking altogether.  The locking is still there, it's just pushed into the Queue implementation.  That's easier for our application, to be sure, but does come with caveats -- for example, we'd no longer be able to lock around multiple actions at once, as we do now.  I don't know whether we really _need_ to do that, but it appears that we _do_ do that now, so we'd need to be mindful of the change.

I'm not familiar with the specifics of how the library maintainers have chosen to implement locking in that component, so I can't really speak to whether the performance characteristics and other behaviors of their choices are suitable for our needs here.

shawkinsmezmo replied:
>I'm not sure the built-in Queue has the right semantics for the use here -- that is, that Queue is more-or-less designed around the notion of each item in the queue being handled discretely from the others, whereas in our use we want to treat everything currently in the queue as a single block of data to reduce the number of separate requests that are sent to our servers. Obviously you could implement such semantics using the Queue but it's not quite as simple as saying "let's just use a queue". In some ways it might be quite clunky to do, in fact, since the Queue would not prohibit threads adding to the queue at just the moment the flushing thread was trying to pull things out to buffer for transmission. Again, not an unsolvable problem but not an obvious slam dunk replacement either, imo.

>Also a minor semantic nit: it's misleading to suggest that using Queue would remove locking altogether. The locking is still there, it's just pushed into the Queue implementation. That's easier for our application, to be sure, but does come with caveats -- for example, we'd no longer be able to lock around multiple actions at once, as we do now. I don't know whether we really need to do that, but it appears that we do do that now, so we'd need to be mindful of the change.

>I'm not familiar with the specifics of how the library maintainers have chosen to implement locking in that component, so I can't really speak to whether the performance characteristics and other behaviors of their choices are suitable for our needs here.

Right. My thought would be that we do something like the following at the time we send data:
```
# Get the size of the queue. Since there is only ever one thread consuming from it this is guaranteed to only either stay the same size or increase during the time this thread is executing
curr_queue_size = self.queue.qsize()

# Consume elements from queue
logs_to_send = [self.queue.get(block=False) for _ in range(curr_queue_size)]

# Continue the flush per existing semantics
self.send_request(logs_to_send)
```

That being said, I see your point and can see how the Queue implementation could possibly yield worse performance due to context switches between threads while handling multiple concurrent calls to `add()` and `get()`. The current implementation has the advantage that the thing that does the draining can consume all of the buffered data in a single locked section.

emelski replied:
A less drastic change from the current design might be to have a single buffer in our class (doing away with `self.secondary`), with very simple locking just around pushing to it and swapping it with a local buffer in `try_request` or one of its callers.  Something like this:

```
def log(message):
    self.lock.acquire(True)
    self.buf.append(message)
    self.lock.release()

def try_request():
    self.lock.acquire(True)
    if not self.buf:
        self.lock.release()
        return

    localbuf = self.buf
    self.buf = []
    self.lock.release()

    // send the stuff in localbuf
```

Obviously this is not a functional replacement, it's just meant to demonstrate the intent -- the locking protocol is vastly simplified, we still have the ability to manage entire buffers of messages at once without any funny business, and we can continue to accumulate new messages while sending the old.  This is basically how the node agent works in fact, so it's not really a novel design even within our code base.

emelski replied:
In case it's not clear, I'm not necessarily advocating for one or the other approach.  Both (or even all three?) suggestions on the table have merit, so as the implementer I think the choice is up to you.  But I do think _something_ needs to be done about the race over `self.secondary`, or the problems will just persist, albeit perhaps at a lower rate than originally observed.

shawkinsmezmo replied:
I like that. I'll take a crack at getting that working.

shawkinsmezmo commented:
@emelski I think this is in line with your suggestion. Removed the secondary buf and implemented the suggested locking semantics. The code is simpler, and I like the idea that there is precedent for this pattern. Based on the testing that I did, this does not introduce any regressions relative to my prior commit.

shawkinsmezmo commented:
force-pushed a small squash commit in order to address linting error

emelski reviewed (commented):
Yeah, I think this looks pretty good!  I noticed a couple remaining issues though:

`self.exception_flag` is written in a thread-unsafe manner now, via `try_request` and `clean_after_success`.  I'd suggest simplifying that code by folding `clean_after_success` into `try_request` since there's not any real gain in having it separate now, and then modifying `try_request` to set a local variable to true/false for the exception state, and adding a  simple block at the end of the function like this:

```
    self.lock.acquire(True)
    self.exception_flag = local_exception_flag
    self.lock.release()
```

Next, the locking in `try_lock_and_do_flush_request` seems to be broken in that it is possible to have the following sequence:

1. Caller calls `buffer_log_sync` which acquires the lock
2. Buffer size after adding the message exceeds `self.flush_limit` so `buffer_log_sync` calls `self.flush()`
3. `self.flush` calls `schedule_flush_sync`
4. `schedule_flush_sync` fails to queue the flush via the request thread pool for some reason
5. `schedule_flush_sync` directly calls `try_lock_and_do_flush_request`
6. `try_lock_and_do_flush_request` fails to get the lock because it is still held in `buffer_log_sync`
7. `buffer_log_sync` fails to flush the request and instead tries to kill and restart the flusher, which doesn't seem desirable.

shawkinsmezmo commented:
@emelski The code as it's currently constructed intends to set this flag to true in the event that the flush request fails, with the intention to communicate that the asynchronously running flush timer thread has been killed. It is inspected by a different thread handling a `buffer_log_sync()` method call (along with other determining factors ie: buffer size), essentially in order to determine whether to:
* Kill the flush timer thread and attempt a flush of buffered data
* (Re)start the flush timer thread
```
def buffer_log_sync(self, message):
    if self.lock.acquire(True):
    try:
        # append to buffer
        
        if self.buf_size >= self.flush_limit and \
                not self.exception_flag:
            self.close_flusher()
            self.flush()
        else:
            self.start_flusher()
    except:
        # log error ...
    finally:
        self.lock.release()
```
I question the value of `exception_flag` and whether it is worthwhile to ever stop / restart the flush timer thread. Especially if it's going to be restarted as soon as another log message is handled. I think it might be cleaner to instead allow that thread to live for the duration of the application's lifecycle, and just ensure with proper locking that calls to flush (acquire lock -> check / consumer buffer -> release lock-> send data if necessary) can be safely invoked from different threads.

shawkinsmezmo commented:
@emelski I pushed a diff that removes the exception_flag altogether, as well as refactors the code so that the flusher thread is started once during application initialization. This removes the need for additional locking, and also utilizes the Timer in a more idiomatic manner.

emelski commented on logdna/logdna.py:
```diff
@@ -99,72 +106,61 @@ def buffer_log(self, message):
-        self.buf.clear()
-        self.buf_size = 0
-        self.exception_flag = False
+                if self.buf_size >= self.flush_limit:
+                    self.flush()
```
I _think_ this is still a small issue, in that this call to `flush()` happens with the lock held, and if `schedule_flush_sync()` is unable to submit the flush to the request thread pool, then it will call `try_lock_and_do_flush_request` itself, which will fail to do anything because the lock is already held.  In theory `submit()` could throw an exception if the pool is shutting down or if it failed to create a new worker thread.  I'm not sure either of those conditions is really worth worrying about, and if the failure mode is just "we don't auto-flush the buffer in that case" that's probably fine -- it's just that the code is a little confused around that case.

I don't feel very strongly about it but if you wanted to address this, you might do something like check the buffer size inside the lock and set a local flag `doFlush = True` as needed; then check that flag and call `flush()` if necessary after releasing the lock.  That would introduce a possible race where multiple flush requests could get queued but that should be harmless as the ultimate flush logic will do nothing if there's nothing to flush.

As I said, I don't feel very strongly about it.  I'm mostly just noting it for posterity I guess.

shawkinsmezmo replied:
>if the failure mode is just "we don't auto-flush the buffer in that case" that's probably fine -- it's just that the code is a little confused around that case.

Yup. The next buffer flush attempt would occur when either the flush timer thread wakes up, or `buffer_log_message()` is invoked, whichever is handled first. The current default configuration value for this thread is `250ms`, but that can be overridden by environment variables. 

Regarding the potential for attempts to submit flush requests to the thread pool failing, the [docs for python's concurrent futures module](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor.shutdown) only state that `RuntimeError` will be raised if `submit()` is invoked after `shutdown()`. The code order in the handler's `close()` method guards against this by ensuring that the executor that is used to handle flush requests can only be shutdown after the executor that handles `buffer_log_message()` requests.  Outside of this, there could be a scenario where there is a system or OS resource issue / limit, but other than that, I think the calling code should always be able to reliably submit the request.  Whether the request succeeds or not will be a different matter.

emelski reviewed (approved):
This looks good to me.  Much simpler and I think the notable races that could result in data loss have been addressed.
